rust/blockstore/src/arrow/block/types.rs

```diff
@@ -118,16 +119,67 @@ impl Block {
         delta
     }
 
+    /// Binary search the block to find the last index of the specified prefix.
+    /// Returns None if prefix does not exist in the block.
+    ///
+    /// Partly based on `std::slice::binary_search_by`: https://doc.rust-lang.org/src/core/slice/mod.rs.html#2770
+    #[inline]
+    fn binary_search_last_index(&self, prefix: &str) -> Option<usize> {
+        let mut size = self.len();
+        if size == 0 {
+            return None;
+        }
+
+        let prefix_array = self
+            .data
+            .column(0)
+            .as_any()
+            .downcast_ref::<StringArray>()
+            .unwrap();
+        let mut base = self.len() - 1;
+
+        // This loop intentionally doesn't have an early exit if the comparison
+        // returns Equal. We want the number of loop iterations to depend *only*
+        // on the size of the input slice so that the CPU can reliably predict
+        // the loop count.
+        while size > 1 {
+            let half = size / 2;
+            let mid = base - half;
+
+            // SAFETY: the call is made safe by the following inconstants:
+            // - `mid < size`: by definition
+            // - `mid >= 0`: `mid = size - 1 - size / 2 - size / 4 ...`
+            let cmp = prefix_array.value(mid).cmp(prefix);
+
+            base = if cmp == Greater { mid } else { base };
+            size -= half;
+        }
+
+        // SAFETY: `base` is always in [0, size) because `base < size` by init.
+        // `base` should be the last index where the element matches the target prefix,
+        // or 0 if the first element is already larger than the target prefix.
+        match prefix_array.value(base).cmp(prefix) {
+            Less => None,
+            Equal => Some(base),
+            Greater => {
+                if base > 0 && prefix_array.value(base - 1) == prefix {
+                    Some(base - 1)
+                } else {
+                    None
+                }
+            }
+        }
+    }
+
     /// Binary search the blockfile to find the partition point of the specified prefix and key.
-    /// The implementation is based on [`std::slice::partition_point`].
     ///
     /// `(prefix, key)` serves as the search key, and it is sorted in ascending order.
     /// The partition predicate is defined by: `|x| x < (prefix, key)`.
     /// The partition point is the first index where the partition precidate evaluates to `false`
     /// The code is a result of inlining this predicate in [`std::slice::partition_point`].
     /// If the key is unspecified (i.e. `None`), we find the first index of the prefix.
     ///
-    /// [`std::slice::partition_point`]: std::slice::partition_point
+    /// Partly based on `std::slice::binary_search_by`: https://doc.rust-lang.org/src/core/slice/mod.rs.html#2770
     #[inline]
     fn binary_search_index<'me, K: ArrowReadableKey<'me>>(
```

Comment:
nit: Both the method names `binary_search_index` and `binary_search_last_index` convey very less in what they do other than that they do binary search. Prefer changing to more meaningful names.